src/payments/service.ts

```diff
@@ -83,6 +87,163 @@ export class PaymentsService {
     return this._paymentsContract
   }
 
+  /**
+   * Generate EIP-2612 permit signature for USDFC token
+   * Handles balance check, domain creation, nonce retrieval, and signature generation
+   * Uses Multicall3 to batch RPC calls for efficiency
+   * @param amount - Amount to permit
+   * @param deadline - Unix timestamp (seconds) when the permit expires
+   * @param contextName - Context name for error messages (e.g., 'depositWithPermit')
+   * @returns Signature object
+   */
+  private async _getPermitSignature(amount: bigint, deadline: bigint, contextName: string): Promise<ethers.Signature> {
+    const signerAddress = await this._signer.getAddress()
+
+    // Get network type (validates network and makes single getNetwork() call internally)
+    const networkType = await getFilecoinNetworkType(this._provider)
+
+    // Derive chainId from network type
+    const chainId = CHAIN_IDS[networkType]
+
+    // Setup Multicall3 for batched RPC calls
+    const multicall3Address = CONTRACT_ADDRESSES.MULTICALL3[networkType]
+    const multicall = new ethers.Contract(multicall3Address, CONTRACT_ABIS.MULTICALL3, this._provider)
+
+    // Create interfaces for encoding/decoding
+    const erc20Interface = new ethers.Interface(CONTRACT_ABIS.ERC20)
+    const permitInterface = new ethers.Interface(CONTRACT_ABIS.ERC20_PERMIT)
+
+    // Prepare multicall batch: balanceOf, name, version (with fallback), nonces
+    const calls = [
+      {
+        target: this._usdfcAddress,
+        allowFailure: false,
+        callData: erc20Interface.encodeFunctionData('balanceOf', [signerAddress]),
+      },
+      {
+        target: this._usdfcAddress,
+        allowFailure: false,
```

Comment:
since you don't allowFailure for some of these, you need to check the length of the results.